airflow/sentry.py

```diff
@@ -0,0 +1,158 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Any
+
+from airflow import configuration as conf
+from airflow.utils.db import provide_session
+from airflow.utils.log.logging_mixin import LoggingMixin
+
+
+log = LoggingMixin().log
+
+
+class DummySentry:
+    """
+    Blank class for Sentry.
+    """
+
+    @classmethod
+    def add_tagging(cls, task_instance):
+        pass
+
+    @classmethod
+    def add_breadcrumbs(cls, session=None):
+        pass
+
+
+@provide_session
+def get_task_instance(task, execution_date, session=None):
+    """
+    Retrieve attribute from task.
+    """
+    from airflow.models.taskinstance import TaskInstance  # Avoid circular import
+
+    if session is None:
+        return None
+    TI = TaskInstance
+    ti = (
+        session.query(TI)
+        .filter(
+            TI.dag_id == task.dag_id,
+            TI.task_id == task.task_id,
+            TI.execution_date == execution_date,
+        )
+        .first()
+    )
+    return ti
+
+
+class ConfiguredSentry:
+    """
+    Configure Sentry SDK.
+    """
+
+    SCOPE_TAGS = frozenset(("task_id", "dag_id", "execution_date", "operator"))
+
+    def __init__(self):
+        """
+        Initialize the Sentry SDK.
+        """
+
+        integrations = []
+        ignore_logger("airflow.task")
+        executor_name = conf.get("core", "EXECUTOR")
+
+        sentry_flask = FlaskIntegration()
+        if executor_name == "CeleryExecutor":
+            from sentry_sdk.integrations.celery import CeleryIntegration
+
+            sentry_celery = CeleryIntegration()
+            integrations = [sentry_celery]
+        else:
+            import logging
+            from sentry_sdk.integrations.logging import LoggingIntegration
+
+            sentry_logging = LoggingIntegration(
+                level=logging.INFO, event_level=logging.ERROR
+            )
+            integrations = [sentry_logging]
+
+        integrations.append(sentry_flask)
+
+        self.task = None
+        self.execution_date = None
+
+        dsn = conf.get("sentry", "sentry_dsn")
+        if dsn:
+            init(dsn=dsn, integrations=integrations)
+        else:
+            # Setting up Sentry using environment variables.
+            init(integrations=integrations)
+
+    def add_tagging(self, task_instance):
+        """
+        Function to add tagging for a task_instance.
+        """
+        self.task = task_instance.task
+        self.execution_date = task_instance.execution_date
+
+        with configure_scope() as scope:
```

Comment:
I am not familiar with sentry internals, but how will exceptions generated by task instances be contained in the sentry scope that you are opening here?

Reply:
You make a good point. I had it such that the tags would overwrite each other, in the Celery case since each execution was run in a different process, tags would not leak. However, for other executors, tags would leak into other exceptions.